src/xdplwf/offload.c

```diff
@@ -1226,13 +1199,257 @@ XdpLwfOffloadRssDeactivate(
 
 static
 NTSTATUS
-XdpLwfOpenInterfaceOffloadHandle(
+XdpLwfOffloadQeoSet(
     _In_ XDP_LWF_FILTER *Filter,
+    _In_ XDP_LWF_INTERFACE_OFFLOAD_CONTEXT *OffloadContext,
+    _In_ const XDP_OFFLOAD_PARAMS_QEO *XdpQeoParams,
+    _In_ UINT32 XdpQeoParamsSize,
+    _Out_opt_ VOID *OffloadResult,
+    _In_ UINT32 OffloadResultSize,
+    _Out_opt_ UINT32 *OffloadResultWritten
+    )
+{
+    NTSTATUS Status;
+    const XDP_QUIC_CONNECTION *XdpQeoConnection;
+    NDIS_QUIC_CONNECTION *NdisConnections = NULL;
+    UINT32 NdisConnectionsSize;
+    ULONG NdisBytesReturned;
+
+    TraceEnter(TRACE_LWF, "Filter=%p OffloadContext=%p", Filter, OffloadContext);
+
+    RtlAcquirePushLockExclusive(&Filter->Offload.Lock);
+
+    if (XdpQeoParamsSize != sizeof(*XdpQeoParams) ||
+        XdpQeoParams->ConnectionCount == 0 ||
+        OffloadResultSize < XdpQeoParams->ConnectionsSize) {
+        Status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+
+    if (OffloadContext->IsInvalid) {
+        TraceError(
+            TRACE_LWF,
+            "OffloadContext=%p Interface offload context invalidated",
+            OffloadContext);
+        Status = STATUS_INVALID_DEVICE_STATE;
+        goto Exit;
+    }
+
+    if (OffloadContext->Edge != XdpOffloadEdgeLower) {
+        Status = STATUS_NOT_SUPPORTED;
+        goto Exit;
+    }
+
+    //
+    // Clone the input params.
+    //
+
+    Status =
+        RtlUInt32Mult(
+            XdpQeoParams->ConnectionCount, sizeof(*NdisConnections), &NdisConnectionsSize);
+    if (!NT_SUCCESS(Status)) {
+        goto Exit;
+    }
+
+    NdisConnections = ExAllocatePoolZero(NonPagedPoolNx, NdisConnectionsSize, POOLTAG_OFFLOAD);
+    if (NdisConnections == NULL) {
+        TraceError(
+            TRACE_LWF, "OffloadContext=%p Failed to allocate NDIS connections", OffloadContext);
+        Status = STATUS_NO_MEMORY;
+        goto Exit;
+    }
+
+    //
+    // Form and issue the OID.
```

Comment:
This is painful. Can we not have a way where a memcpy of the struct is used instead? As mentioned in the other comment, I want this to be extensible, so long as the size of the struct remains constant.

Reply:
I don't think it's unreasonable to have to update XDP if we add a new field/value to this structure. At the very least, NIC drivers and applications will need to be updated. Since XDP is undocked from Windows, I think we should do the right thing here at the user/kernel boundary and validate parameters passed to third-party drivers.